pkg/epp/scheduling/scheduler.go

```diff
@@ -20,141 +20,197 @@ package scheduling
-func (s *Scheduler) Schedule(ctx context.Context, req *types.LLMRequest) (targetPod types.Pod, err error) {
+func (s *Scheduler) Schedule(ctx context.Context, req *types.LLMRequest) (*types.Result, error) {
 	logger := log.FromContext(ctx).WithValues("request", req)
+	loggerDebug := logger.V(logutil.DEBUG)
 
 	// Snapshot pod metrics from the datastore to:
 	// 1. Reduce concurrent access to the datastore.
 	// 2. Ensure consistent data during the scheduling operation of a request.
 	sCtx := types.NewContext(ctx, req, types.ToSchedulerPodMetrics(s.datastore.PodGetAll()))
-	logger.V(logutil.DEBUG).Info(fmt.Sprintf("Scheduling a request. Metrics: %+v", sCtx.PodsSnapshot))
+	loggerDebug.Info(fmt.Sprintf("Scheduling a request. Metrics: %+v", sCtx.PodsSnapshot))
 
-	var filter Filter
-	if req.Critical {
-		filter = s.criticalRequestFilter
-	} else {
-		filter = s.sheddableRequestFilter
+	s.runPreSchedulePlugins(sCtx)
+
+	pods, err := s.runFilterPlugins(sCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.runScorerPlugins(sCtx, pods); err != nil {
+		return nil, err
+	}
+
+	before := time.Now()
+	res, err := s.picker.Pick(sCtx, pods)
+	metrics.RecordSchedulerPluginProcessingLatency(types.PickerPluginType, s.picker.Name(), time.Since(before))
+	if err != nil {
+		return nil, err
 	}
+	loggerDebug.Info("After running picker plugins", "result", res)
 
-	pods, err := filter.Filter(sCtx, sCtx.PodsSnapshot)
-	if err != nil || len(pods) == 0 {
-		return nil, fmt.Errorf("failed to apply filter, resulted %v pods, this should never happen: %w", len(pods), err)
+	s.runPostSchedulePlugins(sCtx, res)
+
+	return res, nil
+}
+
+func (s *Scheduler) runPreSchedulePlugins(ctx *types.Context) {
+	for _, plugin := range s.preSchedulePlugins {
+		ctx.Logger.V(logutil.DEBUG).Info("Running pre-schedule plugin", "plugin", plugin.Name())
+		before := time.Now()
+		plugin.PreSchedule(ctx)
+		metrics.RecordSchedulerPluginProcessingLatency(types.PreSchedulerPluginType, plugin.Name(), time.Since(before))
+	}
+}
+
+func (s *Scheduler) runPostSchedulePlugins(ctx *types.Context, res *types.Result) {
+	for _, plugin := range s.postSchedulePlugins {
+		ctx.Logger.V(logutil.DEBUG).Info("Running post-schedule plugin", "plugin", plugin.Name())
+		before := time.Now()
+		plugin.PostSchedule(ctx, res)
+		metrics.RecordSchedulerPluginProcessingLatency(types.PostSchedulePluginType, plugin.Name(), time.Since(before))
+	}
+}
+
+func (s *Scheduler) runFilterPlugins(ctx *types.Context) ([]types.Pod, error) {
+	loggerDebug := ctx.Logger.V(logutil.DEBUG)
+	pods := ctx.PodsSnapshot
+	loggerDebug.Info("Before running filter plugins", "pods", pods)
+	for _, filter := range s.filters {
+		loggerDebug.Info("Running filter plugin", "plugin", filter.Name())
+		before := time.Now()
+		filteredPods, err := filter.Filter(ctx, pods)
+		metrics.RecordSchedulerPluginProcessingLatency(types.FilterPluginType, filter.Name(), time.Since(before))
```

Comment:
this is not measurement of each filter. since the filter is defined as a decision tree, these lines measures the decision tree time, means it calculates all filters.
this is another reason for transitioning to filter chain and not decision tree (when removing error from filter as noted in the other comment). we should be able to see how much time each filter took.